compiler/rustc_resolve/src/lib.rs

```diff
@@ -1023,6 +1025,8 @@ pub struct Resolver<'a> {
     trait_impl_items: FxHashSet<LocalDefId>,
 
     legacy_const_generic_args: FxHashMap<DefId, Option<Vec<usize>>>,
+
+    main_by_path: Option<(DefId, DefKind)>,
```

Comment:
```suggestion
    main_by_path: Option<Res>,
```
If you are reporting the "something at `crate::main`, but it is not a function" later in type checking, then it makes sense to keep all resolutions.

Reply:
Stored `Res` and span and other things from NameBinding into a new struct.